src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -0,0 +1,329 @@
+            FAIL_FAST_IF_NULL(cmd);
+
+            return !cmd.HasNestedCommands() && cmd.Action().Action() == ShortcutAction::Invalid ?
+                       nullptr : // explicitly unbound
+                       cmd;
+        }
+    }
+
+    // Method Description:
+    // - Retrieves a map of command names to the commands themselves
+    // - These commands should not be modified directly because they may result in
+    //    an invalid state for the `ActionMap`
+    Windows::Foundation::Collections::IMapView<hstring, Model::Command> ActionMap::NameMap()
+    {
+        if (_NameMapCache)
+        {
+            return _NameMapCache.GetView();
+        }
+
+        std::unordered_map<hstring, Model::Command> nameMap{};
+        std::set<size_t> visitedActionIDs{ ActionHash()(make<implementation::ActionAndArgs>()) };
+        _PopulateNameMap(nameMap, visitedActionIDs);
+
+        _NameMapCache = single_threaded_map<hstring, Model::Command>(std::move(nameMap));
+        return _NameMapCache.GetView();
+    }
+
+    // Method Description:
+    // - Populates the provided nameMap with all of our actions and our parents actions
+    //    while omitting the actions that were already added before
+    // Arguments:
+    // - nameMap: the nameMap we're populating. This maps the name (hstring) of a command to the command itself.
+    //             There should only ever by one of each command (identified by the actionID) in the nameMap.
+    // - visitedActionIDs: the actionIDs that we've already added to the nameMap. Commands with a matching actionID
+    //                      have already been added, and should be ignored.
+    void ActionMap::_PopulateNameMap(std::unordered_map<hstring, Model::Command>& nameMap, std::set<size_t>& visitedActionIDs) const
+    {
+        // Update NameMap and visitedActionIDs with our current layer
+        for (const auto& [actionID, cmd] : _ActionMap)
+        {
+            // Only populate NameMap with actions that haven't been visited already.
+            if (visitedActionIDs.find(actionID) == visitedActionIDs.end())
+            {
+                const auto& name{ cmd.Name() };
+                if (!name.empty())
+                {
+                    // Update NameMap.
+                    nameMap.insert_or_assign(name, cmd);
+                }
+
+                // Record that we already handled adding this action to the NameMap.
+                visitedActionIDs.insert(actionID);
+            }
+        }
+
+        // Update NameMap and visitedActionIDs with our parents
+        for (const auto& parent : _parents)
+        {
+            parent->_PopulateNameMap(nameMap, visitedActionIDs);
+        }
+
+        // Add NestedCommands to NameMap _after_ we handle our parents.
+        // This allows us to override whatever our parents tell us.
+        for (const auto& [name, cmd] : _NestedCommands)
+        {
+            nameMap.insert_or_assign(name, cmd);
+        }

```

Comment:
Yeah, this one's trippy. Nested commands are special in that we have no way to check for equality. Thus, we can't add them to `visitedActionIDs`. Instead, we do the following:
1. In `AddAction`, remove/update `_NestedCommands` if there is a collision (i.e. `copy` action was explicitly set to have the same name as a nested command)
2. In `_PopulateNameMap`...
   a. add nested commands _last_ so that we override any collisions from our parents

Using this method, we make sure that we _do not_ modify our parents. This lets you see the state of our `ActionMap` after loading a specific settings file as opposed to the whole thing.